pkg/xrd/gen.go

```diff
@@ -0,0 +1,318 @@
+		if pkg != nil {
+			return pkg
+		}
+	}
+	return nil
+}
+
+// FindKubeKinds locates all types that contain TypeMeta and ObjectMeta
+// (and thus may be a Kubernetes object), and returns the corresponding
+// group-kinds.
+func FindKubeKinds(parser *crd.Parser, metav1Pkg *loader.Package) []schema.GroupKind {
+	// TODO(directxman12): technically, we should be finding metav1 per-package
+	kubeKinds := map[schema.GroupKind]struct{}{}
+	for typeIdent, info := range parser.Types {
+		hasObjectMeta := false
+		hasTypeMeta := false
+
+		pkg := typeIdent.Package
+		pkg.NeedTypesInfo()
+		typesInfo := pkg.TypesInfo
+
+		for _, field := range info.Fields {
+			if field.Name != "" {
+				// type and object meta are embedded,
+				// so they can't be this
+				continue
+			}
+
+			fieldType := typesInfo.TypeOf(field.RawField.Type)
+			namedField, isNamed := fieldType.(*types.Named)
+			if !isNamed {
+				// ObjectMeta and TypeMeta are named types
+				continue
+			}
+			if namedField.Obj().Pkg() == nil {
+				// Embedded non-builtin universe type (specifically, it's probably `error`),
+				// so it can't be ObjectMeta or TypeMeta
+				continue
+			}
+			fieldPkgPath := loader.NonVendorPath(namedField.Obj().Pkg().Path())
+			fieldPkg := pkg.Imports()[fieldPkgPath]
+
+			// Compare the metav1 package by ID and not by the actual instance
+			// of the object. The objects in memory could be different due to
+			// loading from different root paths, even when they both refer to
+			// the same metav1 package.
+			if fieldPkg == nil || fieldPkg.ID != metav1Pkg.ID {
+				continue
+			}
+
+			switch namedField.Obj().Name() {
+			case "ObjectMeta":
+				hasObjectMeta = true
+			case "TypeMeta":
+				hasTypeMeta = true
+			}
+		}
+
+		if !hasObjectMeta || !hasTypeMeta {
+			continue
+		}
+
+		groupKind := schema.GroupKind{
+			Group: parser.GroupVersions[pkg].Group,
+			Kind:  typeIdent.Name,
+		}
+		kubeKinds[groupKind] = struct{}{}
+	}
+
+	groupKindList := make([]schema.GroupKind, 0, len(kubeKinds))
+	for groupKind := range kubeKinds {
+		groupKindList = append(groupKindList, groupKind)
+	}
+	sort.Slice(groupKindList, func(i, j int) bool {
+		return groupKindList[i].String() < groupKindList[j].String()
+	})
+
+	return groupKindList
+}
+
+// filterTypesForCRDs filters out all nodes that aren't used in CRD generation,
+// like interfaces and struct fields without JSON tag.
+func filterTypesForCRDs(node ast.Node) bool {
+	switch node := node.(type) {
+	case *ast.InterfaceType:
+		// skip interfaces, we never care about references in them
+		return false
+	case *ast.StructType:
+		return true
+	case *ast.Field:
+		_, hasTag := loader.ParseAstTag(node.Tag).Lookup("json")
+		// fields without JSON tags mean we have custom serialization,
+		// so only visit fields with tags.
+		return hasTag
+	default:
+		return true
+	}
+}
+
+// transformRemoveCRDStatus ensures we do not write the CRD status field.
+func transformRemoveXRDStatus(obj map[string]interface{}) error {
```

Comment:
```suggestion
// removeXRDStatus ensures we do not write the CRD status field.
func removeXRDStatus(obj map[string]interface{}) error {
```

Reply:
This was matching existing style from the packages.

Reply:
Strange.   Yeah, if its existing convention we can stick with it.